sdks/go/container/boot.go

```diff
@@ -0,0 +1,127 @@
+
+package main
+
+import (
+	"context"
+	"flag"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/artifact"
+	"github.com/apache/beam/sdks/go/pkg/beam/provision"
+	"github.com/apache/beam/sdks/go/pkg/beam/util/execx"
+	"github.com/apache/beam/sdks/go/pkg/beam/util/grpcx"
+)
+
+var (
+	// Contract: https://s.apache.org/beam-fn-api-container-contract.
+
+	id                = flag.String("id", "", "Local identifier (required).")
+	loggingEndpoint   = flag.String("logging_endpoint", "", "Local logging endpoint for FnHarness (required).")
+	artifactEndpoint  = flag.String("artifact_endpoint", "", "Local artifact endpoint for FnHarness (required).")
+	provisionEndpoint = flag.String("provision_endpoint", "", "Local provision endpoint for FnHarness (required).")
+	controlEndpoint   = flag.String("control_endpoint", "", "Local control endpoint for FnHarness (required).")
+	semiPersistDir    = flag.String("semi_persist_dir", "/tmp", "Local semi-persistent directory (optional).")
+)
+
+func main() {
+	flag.Parse()
+	if *id == "" {
+		log.Fatal("No id provided.")
+	}
+	if *loggingEndpoint == "" {
+		log.Fatal("No logging endpoint provided.")
+	}
+	if *artifactEndpoint == "" {
+		log.Fatal("No artifact endpoint provided.")
+	}
+	if *provisionEndpoint == "" {
+		log.Fatal("No provision endpoint provided.")
+	}
+	if *controlEndpoint == "" {
+		log.Fatal("No control endpoint provided.")
+	}
+
+	log.Printf("Initializing Go harness: %v", strings.Join(os.Args, " "))
+
+	ctx := grpcx.WriteWorkerID(context.Background(), *id)
+
+	// (1) Obtain the pipeline options
+
+	info, err := provision.Info(ctx, *provisionEndpoint)
+	if err != nil {
+		log.Fatalf("Failed to obtain provisioning information: %v", err)
+	}
+	options, err := provision.ProtoToJSON(info.GetPipelineOptions())
+	if err != nil {
+		log.Fatalf("Failed to convert pipeline options: %v", err)
+	}
+
+	// (2) Retrieve the staged files.
+	//
+	// The Go SDK harness downloads the worker binary and invokes
+	// it. For now, we assume that the first (and only) package
+	// is the binary.
+
+	dir := filepath.Join(*semiPersistDir, "staged")
+	artifacts, err := artifact.Materialize(ctx, *artifactEndpoint, dir)
+	if err != nil {
+		log.Fatalf("Failed to retrieve staged files: %v", err)
+	}
+	if len(artifacts) == 0 {
+		log.Fatal("No binaries staged")
+	}
+
+	// (3) The persist dir may be on a noexec volume, so we must
+	// copy the binary to a different location to execute.
+
+	prog := filepath.Join("/bin", artifacts[0].Name)
+	if err := copyExe(filepath.Join(dir, artifacts[0].Name), prog); err != nil {
+		log.Fatalf("Failed to copy binary: %v", err)
+	}
+
+	args := []string{
+		"--worker=true",
+		"--id=" + *id,
+		"--logging_endpoint=" + *loggingEndpoint,
+		"--control_endpoint=" + *controlEndpoint,
+		"--semi_persist_dir=" + *semiPersistDir,
+		"--options=" + options,
+	}
+	log.Fatalf("User program exited: %v", execx.Execute(prog, args...))
```

Comment:
<!--thread_id:cc_155324829_t; commit:62de5ca50edc57ee9242dc3ce901ab9995f7238b; resolved:1-->
<!--section:context-quote-->
> **lostluck** wrote:
> If the program exited, this fatal log runs, leading stopping the program from... exiting. I assume then the user program is never expected to exit?
> 
> If that's an expected precondition (the given user program never terminates), then please add a comment to that effect.
> 
> If not, please change the log to Printf instead.

<!--section:body-->
No, it simply logs the error from the user program and then exits itself. It's a standard idiom and I don't think a comment is necessary. A user program exit becomes a container exit. The user program should not exit under normal operation, but it may. In that case, the program and container exit and are restarted by the runner.

Reply:
Btw, this code is also how the other SDK container images look.